### ISSUE TITLE
test(perf): VCS fixtures spawn ~680 git processes per run

### ISSUE THREAD

**Issue by dekobon** [dekobon/big-code-analysis#1125]
## Summary

The VCS test fixture builder shells out to `git` per operation: `Repo::init()` (`tests/common/vcs_fixture.rs:28-51`) spawns **5** processes (`git init` + 4 `git config`; the hooks dir is a plain `create_dir_all`), and `commit_inner` (`:100-125`) spawns 2 more per commit (`add -A`, `commit`). `run_git` (`:129-137`) is a bare `Command::new("git")…status()`.

**Measured rather than estimated.** Shimming `git` on `PATH` with a counter and running the seven prebuilt VCS test binaries:

| Binary | `#[test]` fns | git spawns | wall (8 threads) |
|---|---|---|---|
| `vcs_history` | 32 | 291 | 0.63 s |
| `vcs_jit` | 18 | 169 | 0.27 s |
| `vcs_per_function` | 7 | **477** | **7.98 s** |
| `vcs_cache` | 13 | 126 | 0.43 s |
| `vcs_trend` | 9 | 84 | 0.26 s |
| `vcs_bus_factor` | 7 | 61 | 0.09 s |
| `vcs_file_types` | 6 | 44 | 0.10 s |
| **total** | **92** | **1 252** | **~9.8 s** |

Subcommand histogram: `commit` 391, `add` 391, `config` 364, `init` 91, `mv` 5, `checkout` 4, `rm` 3, `merge` 2, `clone` 1. (91 inits × 4 configs = 364 confirms the 5-spawn init exactly.)

So **1 252 spawns, ~1.8× the original ~680 estimate** — but the per-spawn cost was overestimated in the other direction, and that matters more. Micro-benchmarked here: 100 `init`/`config` spawns in 0.186 s (**~1.9 ms/spawn**, ~9.3 ms per `Repo::init()`); 40 `add`/`commit` spawns in 0.171 s (~4.3 ms/spawn). Not 5-15 ms.

**Consequence: process churn is ~3-5 s of the ~9.8 s, and it is not where the wall time concentrates.** `perf_10k_lines_200_commits_under_30s` (`tests/vcs_per_function.rs:226-327`) alone accounts for **407 of `vcs_per_function`'s 477 spawns** and 13.4-23.7 s when run alone; that binary's other six tests together cost 70 spawns / 0.066 s. The remaining six binaries cost **1.68 s combined**, which is the hard ceiling on what fixture sharing can save.

## Resolution Plan

Reordered by measured payoff.

1. **Shrink `perf_10k_lines_200_commits_under_30s` — the single biggest win.** Its fixture makes 201 commits, each re-rendering and rewriting a full 10 000-line file; the `format!` loop and file writes rival the subprocess cost. The 200/10k figures exist to satisfy a 30 s acceptance budget it currently clears by ~2×. Cutting either buys more than every other item here combined. It also cannot be shared (it mutates), so items 2-3 do not reach it.
2. **Port the fixture to in-process `gix`** — this was step 3 ("larger change, only if 1-2 prove insufficient"). That framing is wrong: it is *smaller* than assumed, and there is a working reference implementation in [`dekobon/git-remote-object-store`](https://github.com/dekobon/git-remote-object-store), whose unit tests build repos entirely in-process while only its end-to-end tier shells out. Pattern (`src/git/branch.rs:36-80`):

   ```rust
   let repo = gix::init(dir.path())?;                    // no config needed
   let blob = repo.write_blob(bytes)?.detach();
   let tree = repo.write_object(&gix::objs::Tree { entries })?.detach();
   repo.commit_as(sig, sig, "refs/heads/main", msg, tree, parents)?.detach()
   ```
   where `sig` is `gix::actor::SignatureRef { name, email, time: "0 +0000" }`.

   Why this is a better fit than batching the config calls:

   - **All four `git config` calls become structurally unnecessary, not merely batched.** `commit.gpgsign`, `core.hooksPath`, `gc.auto`, and `core.autocrlf` exist solely to neutralise host config. `commit_as` does not sign, does not run hooks, does not gc, and writes blobs from bytes we hand it — no smudge/clean filter. The entire class of host-contamination bug that #941 patched becomes unrepresentable. That removes 364 config spawns *and* the reason they exist.
   - **`SignatureRef::time` is a raw git wire-format `&str`** — the same `"@{secs} +0000"` shape `commit_inner` already passes via `GIT_AUTHOR_DATE`/`GIT_COMMITTER_DATE`. Our fixture already supplies name, email, and time explicitly per commit, so the port is close to mechanical, and it yields deterministic SHAs.
   - **`commit_as` updates the ref itself** when passed `refs/heads/main` — no separate `edit_reference`.
   - It removes the same-second-hash-collision hazard their subprocess tier had to work around (`src/test_util.rs:275-282`).

   Gotchas to carry over: `.detach()` to get an owned `ObjectId` off the borrowing `gix::Id<'repo>`; **write real trees — `repo.empty_tree()` builds a `Tree` value without writing it, leaving commits pointing at a dangling tree id** (`src/git.rs:1213-1216`); and **sort tree entries lexicographically yourself, gix does not re-sort** (`src/packchain/git.rs:260-261`) — `"inner"` before `"main.rs"`, `"Cargo.toml"` before `"src"`.

   Scope limits, honestly: `add -A` → explicit tree construction is the one real porting cost (the fixture stages whatever is in the worktree). Files must still be `fs::write`-ten to disk as well as blob-written, since the analyzer parses the worktree copy. `.mailmap` is unaffected — it is read from the worktree, no commit needed. The 15 spawns for `merge` (2), `mv` (5), `rm` (3), `checkout` (4) and the single `clone` (`vcs_history.rs:634`, `shallow_clone_degrades_gracefully`) are the awkward tail; keep a subprocess tier for those rather than reimplementing merge.
3. **Share immutable fixtures** across the six non-`vcs_per_function` binaries via `LazyLock<Repo>` statics (there is currently no `LazyLock`/`OnceLock` anywhere in the VCS fixtures — every `Repo` is a fresh `TempDir`). Capped at 1.68 s of savings, and cheaper to do after item 2 makes construction nearly free. Depends on the binary merge in #1124.
4. Confirm `vcs_history` / `vcs_jit` / `vcs_per_function` wall times drop. Note that #1117 will **not** move these numbers — see the attribution note there.

## Assessment

| Dimension  | Rating |
|------------|--------|
| Difficulty | Medium |
| Complexity | Low |
| Priority   | Medium |

---

## Decision (2026-07-31)

**Approved to shrink `perf_10k_lines_200_commits_under_30s`, with the budget
re-derived rather than left as-is.**

The 200-commit / 10 000-line figures exist to satisfy a 30 s acceptance budget
that the test currently clears by roughly 2×, and item 1 is correctly
identified as buying more than every other item here combined.

The condition: **shrink the fixture and the budget together.** Leaving a 30 s
budget over a fixture half the size silently doubles the regression the test
will tolerate — the assertion keeps passing while covering less, which is the
same failure shape the fixture reduction is meant to avoid. Re-derive the
budget from the new fixture's measured time with the same headroom ratio the
current one has, and record the derivation in a comment so the next person
shrinking it does not have to re-measure from scratch.

Items 2 and 3 (sharing immutable fixtures) are unchanged and do not reach this
test, since it mutates.

**Comment by dekobon** [comment]
Rewrote the body after measuring the fixture cost directly and studying the test-fixture strategy in [`dekobon/git-remote-object-store`](https://github.com/dekobon/git-remote-object-store).

**Both numbers in the original summary were wrong, in opposite directions.** Instrumented by shimming `git` on `PATH` with a counter: **1 252 spawns**, not ~680 (`Repo::init` is 5 spawns, not 6 — the hooks dir is `create_dir_all`). But a spawn costs **~1.9 ms**, not 5-15 ms. Net: process churn is ~3-5 s of the suite's ~9.8 s.

**The bigger correction is attribution.** `perf_10k_lines_200_commits_under_30s` accounts for 407 of `vcs_per_function`'s 477 spawns and essentially all of that binary's 7.98 s; the other six tests in it run in 0.066 s, and the other six binaries cost 1.68 s *combined*. So step 2 as originally written (shared `LazyLock` fixtures) has a hard ceiling of 1.68 s and cannot touch the perf test at all, since that test mutates. Shrinking the perf test's 201 commits × 10 000-line rewrites is now step 1.

**Step 3 — "replace the fixture builder with in-process gix, larger change, only if 1-2 prove insufficient" — was backwards, and is now step 2.** `git-remote-object-store` runs a deliberate two-tier split: unit tests build repos in-process with gix, only end-to-end tests shell out. Production there spawns no `git` at all.

The reason it is smaller than it looks: **`commit_as` takes author and committer as explicit `SignatureRef`s, so there is no git config to neutralise.** All four of our `git config` calls (`commit.gpgsign`, `core.hooksPath`, `gc.auto`, `core.autocrlf`) exist only to defend against host config — and in-process none of those mechanisms can fire. `commit_as` does not sign, does not run hooks, does not gc, and blobs are written from bytes we control so no autocrlf filter applies. The #941 class of bug becomes unrepresentable rather than patched. And `SignatureRef::time` is a raw wire-format `&str` (`"0 +0000"`) — the same shape `commit_inner` already passes through `GIT_AUTHOR_DATE`, so our fixture is already supplying exactly the inputs the in-process API wants.

Three gotchas from their code that would each have cost a debugging session, all in the body:

- `repo.empty_tree()` builds a `Tree` value **without writing it** — commits then reference a dangling tree id (`src/git.rs:1213-1216`).
- **gix does not sort tree entries**; you must emit them in git's lexicographic canonical order yourself or you produce a tree that hashes differently than git would (`src/packchain/git.rs:260-261`).
- `write_blob`/`write_object`/`commit_as` return a repo-borrowing `gix::Id<'repo>`; `.detach()` for an owned `ObjectId` you can store.

Plus: `commit_as` updates the ref when given `refs/heads/main` (no `edit_reference` needed — there are zero such calls in their repo), and fixed `time: "0 +0000"` gives deterministic SHAs, which removes the same-second-hash-collision workaround their subprocess tier needed (`src/test_util.rs:275-282`).

Scoped honestly in the body: `add -A` → explicit tree building is the real porting cost, files still need `fs::write` alongside the blob write since the analyzer parses the worktree copy, and the 15 `merge`/`mv`/`rm`/`checkout` spawns plus the one `clone` should stay on a subprocess tier rather than being reimplemented.

One more idea worth stealing for the perf test if it stays large: their 5 000-dir × 16-file scale fixture writes **one** blob and reuses its OID for all 80 000 tree entries (`src/packchain/git.rs:893-927`), building sub-second. And their perf guard asserts on the *work product* (`index.tree.len() == dirs`) alongside the timing, explicitly to catch "a regression that produces an empty path-index without changing the timing" — a timing assertion that does not verify the work happened is worthless.

**Comment by dekobon** [comment]
**Decision recorded**: approved to shrink `perf_10k_lines_200_commits_under_30s` — with the budget re-derived, not left as-is.

Item 1 is correctly identified as buying more than everything else here combined. The condition is that the fixture and the budget shrink **together**: leaving a 30 s budget over a half-size fixture silently doubles the regression the test tolerates, so the assertion keeps passing while covering less — the same failure shape the reduction is meant to avoid. Re-derive from the new fixture's measured time at the same headroom ratio, and record the derivation in a comment.

Items 2 and 3 unchanged; they do not reach this test, since it mutates.

**Comment by dekobon** [comment]
## Item 1 done: fixture and budget shrunk together

`perf_10k_lines_200_commits_under_30s` → `perf_10k_lines_50_commits_under_8s`.
History depth drops 200 → 50 commits; the file stays 10 000 lines / 2 000
spans, since that is the dimension per-function bucketing scales in.

### Budget derivation (now recorded in the test's comment)

Measured on one 16-core host under concurrent load (load average 20-26),
debug build, median of 5 runs, min in brackets. Only the blame call is
timed, so the fixture cost never entered the budget:

| Fixture | blame (timed) | `git` spawns | binary CPU | budget | headroom |
|---|---|---|---|---|---|
| 2 000 funcs / 200 commits | 0.53 s [0.49] | 407 | 1.63 CPU-s | 30 s | 57x |
| 2 000 funcs / 50 commits | 0.14 s [0.13] | 107 | 0.82 CPU-s | 8 s | 57x |

The headroom ratio is what carries over, not the 30 s literal.

The comment also records what the shrink trades away: sensitivity to a
*superlinear*-in-history regression. Per-commit budget is preserved
(30 s / 200 and 8 s / 50 are both ~0.15 s per commit), so a linear
regression trips at the same threshold, but a quadratic one now has to be
~4x worse.

### Work product is now asserted, not just the timing

The old guard was `stats.iter().any(|s| s.commits_long > 0)` — true if a
single span attributed anything. Both windows are exact by construction,
so the test now pins the exact history rather than "non-zero":

- `commits_long` is exactly 2 for the edited functions (seed + their own
  edit) and exactly 1 for the other 1 950. `>= 1` would have admitted a
  regression that credited every function with the whole file's history.
- The set of functions carrying a *recent* commit equals exactly the
  edited set, as an ordered `Vec<usize>` equality.

Verified discriminating by perturbation: shifting the fixture's edit
target by one function (`commit * STRIDE + 1`) fails both assertions
(`100 function(s) have an unexpected commits_long; first few (index,
value): [(7, 1), (8, 2), …]`). The old `any(...)` guard passes it.

### Measurements

`vcs_per_function` binary, 5 runs each, same host, same session:

| | min wall | CPU (user+sys) | `git` spawns |
|---|---|---|---|
| before | 1.69 s | 1.63 s | 499 |
| after | 0.82 s | 0.82 s | 199 |

Across all seven VCS binaries: 1 105 → 805 spawns.

### Corrections to the measurements in this issue

Re-measured rather than trusted, per the issue's own standard.

1. **The 30 s budget was never cleared "by ~2x" — it was cleared by
   ~57x.** The 13.4-23.7 s figure is whole-test wall time including
   fixture setup; the assertion times only `per_function`, which measured
   0.49-0.59 s at the *old* 200-commit size on this host. That
   misattribution is what made the budget look tight. The fixture reduction
   is still right — it just buys setup time, not assertion headroom.
2. **The per-binary spawn counts no longer reproduce.** Measured with the
   same PATH-shim technique: `vcs_jit` 60 (issue: 169), `vcs_trend` 24
   (issue: 84), `vcs_history` 291 and `vcs_cache` 126 (both matching).
   Total 1 105 before this change, not 1 252. `vcs_per_function` measured
   499, not 477 — #1117 added a test to it since.
3. **Wall times in the table are not reproducible on a loaded host at
   all.** The same unmodified `vcs_history` binary measured 0.11 s and
   0.96 s minutes apart. CPU time and per-run minima are the only usable
   comparison here; the tables above use those.

### Items 2 and 3 not attempted

Item 2 (in-process `gix` fixture builder) is left open deliberately. After
item 1 the seven VCS binaries cost ~1.5 s wall combined, so the port's
ceiling is a fraction of that, against rewriting a fixture builder shared
by 92 tests — including `add -A` over nested directories, `git mv`,
`merge --no-ff`, and a `clone`, each of which needs an explicit
tree-construction or subprocess-tier answer. Worth doing on its own merits
(it makes the four `git config` calls structurally unnecessary rather than
merely batched), not as a performance fix. Item 3 still depends on #1124.
